RecoEgamma/EgammaTools/plugins/CalibratedElectronProducers.cc

```diff
@@ -0,0 +1,198 @@
+
+#include <vector>
+
+
+template<typename T>
+class CalibratedElectronProducerT: public edm::stream::EDProducer<>
+{
+public:
+  explicit CalibratedElectronProducerT( const edm::ParameterSet & ) ;
+  ~CalibratedElectronProducerT() override{}
+  void produce( edm::Event &, const edm::EventSetup & ) override ;
+  
+private: 
+  void setSemiDetRandomSeed(const edm::Event& iEvent,const T& obj,size_t nrObjs,size_t objNr);
+
+  edm::EDGetTokenT<edm::View<T> > electronToken_;
+  
+  EpCombinationTool        epCombinationTool_;
+  ElectronEnergyCalibrator energyCorrector_;
+  std::unique_ptr<TRandom> semiDeterministicRng_;
+  edm::EDGetTokenT<EcalRecHitCollection> recHitCollectionEBToken_;
+  edm::EDGetTokenT<EcalRecHitCollection> recHitCollectionEEToken_;
+  bool produceCalibratedObjs_;
+  static const std::vector<std::pair<size_t,std::string> > valMapsToStore_;
+
+  typedef edm::ValueMap<float>                     floatMap;
+  typedef edm::View<T>                             objectCollection;
+  typedef std::vector<T>                           objectVector;
+
+};
+
+template<typename T>
+const std::vector<std::pair<size_t,std::string> > CalibratedElectronProducerT<T>::valMapsToStore_ = {
+  {EGEnergySysIndex::kScaleUp,"EGMscaleUpUncertainty"},
+  {EGEnergySysIndex::kScaleDown,"EGMscaleDownUncertainty"},
+  {EGEnergySysIndex::kScaleStatUp,"EGMscaleStatUpUncertainty"},
+  {EGEnergySysIndex::kScaleStatDown,"EGMscaleStatDownUncertainty"},
+  {EGEnergySysIndex::kScaleSystUp,"EGMscaleSystUpUncertainty"},
+  {EGEnergySysIndex::kScaleSystDown,"EGMscaleSystDownUncertainty"},
+  {EGEnergySysIndex::kScaleGainUp,"EGMscaleGainUpUncertainty"},
+  {EGEnergySysIndex::kScaleGainDown,"EGMscaleGainDownUncertainty"},
+  {EGEnergySysIndex::kSmearUp,"EGMresolutionUpUncertainty"},
+  {EGEnergySysIndex::kSmearDown,"EGMresolutionDownUncertainty"},
+  {EGEnergySysIndex::kSmearRhoUp,"EGMresolutionRhoUpUncertainty"},
+  {EGEnergySysIndex::kSmearRhoDown,"EGMresolutionRhoDownUncertainty"},
+  {EGEnergySysIndex::kSmearPhiUp,"EGMresolutionPhiUpUncertainty"},
+  {EGEnergySysIndex::kSmearPhiDown,"EGMresolutionPhiDownUncertainty"},
+  {EGEnergySysIndex::kScaleValue,"EGMscale"},
+  {EGEnergySysIndex::kSmearValue,"EGMsmear"},
+  {EGEnergySysIndex::kSmearNrSigma,"EGMsmearNrSigma"},
+  {EGEnergySysIndex::kEcalPreCorr,"EGMecalEnergyPreCorr"},
+  {EGEnergySysIndex::kEcalErrPreCorr,"EGMecalEnergyErrPreCorr"},
+  {EGEnergySysIndex::kEcalPostCorr,"EGMecalEnergy"},
+  {EGEnergySysIndex::kEcalErrPostCorr,"EGMecalEnergyErr"},
+  {EGEnergySysIndex::kEcalTrkPreCorr,"EGMecalTrkEnergyPreCorr"},
+  {EGEnergySysIndex::kEcalTrkErrPreCorr,"EGMecalTrkEnergyErrPreCorr"}, 
+  {EGEnergySysIndex::kEcalTrkPostCorr,"EGMecalTrkEnergy"},
+  {EGEnergySysIndex::kEcalTrkErrPostCorr,"EGMecalTrkEnergyErr"}
+};
+
+namespace{
+  template<typename HandleType,typename ValType>
+    void fillAndStoreValueMap(edm::Event& iEvent,HandleType objHandle,
+			      std::vector<ValType> vals,std::string name)
```

Comment:
ah yes, not sure what I was thinking. done

Reply:
btw on this, I feel that there a bunch of utility functions like this that should be defined somewhere. Otherwise everybody just makes their own every time. Are you aware of anything like these or could suggest an appropriate area to put utility tools such as these.  This and getHandle<T>() are something I end up always defining.